topdown/eval.go

```diff
@@ -1682,30 +1685,65 @@ func (e evalBuiltin) eval(iter unifyIterator) error {
 	numDeclArgs := len(e.bi.Decl.FuncArgs().Args)
 
 	e.e.instr.startTimer(evalOpBuiltinCall)
+	var err error
+	needNormalUnify := true
+	endIndex := len(e.terms) - 1
 
-	err := e.f(e.bctx, operands, func(output *ast.Term) error {
-
+	// We skip evaluation of the builtin entirely if the NDBuiltinCache is
+	// present, and we have a non-deterministic builtin already cached.
+	if e.bi.Nondeterministic && e.bctx.NDBuiltinCache != nil {
 		e.e.instr.stopTimer(evalOpBuiltinCall)
 
-		var err error
-
-		if e.bi.Decl.Result() == nil {
-			err = iter()
-		} else if len(operands) == numDeclArgs {
-			if output.Value.Compare(ast.Boolean(false)) != 0 {
-				err = iter()
-			}
-		} else {
-			err = e.e.unify(e.terms[len(e.terms)-1], output, iter)
+		// NOTE(philipc): We have to drop the very last term off the args list,
+		// as it's a generated term, not an actual argument to the builtin.
```

Comment:
So, before (and now below), we're figuring this out,
```go
} else if len(operands) == numDeclArgs {
```
I think this may or may not be the case. With how the code stands here, we assume that it's always the case.

Here's an example:

```rego
package x

p {
	http.send({"url": "http://httpbin.org", "method": "GET"})
}
```
When checking how the compiler rewrites this, there's _no_ extra argument: 
```rego
p = true { http.send({"method": "GET", "url": "http://httpbin.org"}) }]
```

It's an edge case, of course, as anything using the response would be rewritten, like
```rego
p {
	http.send({"url": "http://httpbin.org", "method": "GET"}).body == {"foo": "bar"}
}
```
⏬
```rego
p = true { http.send({"method": "GET", "url": "http://httpbin.org"}, __local0__); __local0__.body = {"foo": "bar"} }]
```

Reply:
Would adding a guard to check for that case be sufficient protection? I've added more guard logic to do that in my recent changes, but if you think there's more that should be done, let me know.

Reply:
Yeah I think we need something along these lines,
```go
endIndex := len(operands)
if len(operands) > numDeclArgs {
  endIndex--
}
```
because with what we have right now,
```go
if e.bi.Decl.Result() != nil && len(operands) != numDeclArgs {
```
would we not ignore the NDCache in the http.send example above? 

Perhaps a test would help. The ones we have use queries, and not built-in calls within bodies; I think that could mislead is here, since those should always end up with captured output vars.

Reply:
```go
func TestNBCache(t *testing.T) {
	ctx := context.Background()
	ndBC := builtins.NDBCache{}
	query := "data.foo.p = x"
	_, err := New(
		Query(query),
		NDBuiltinCache(ndBC),
		Module("test.rego", `package foo
p {
	http.send({"url": "http://httpbin.org", "method":"get"})
}`),
	).Eval(ctx)
	if err != nil {
		t.Fatal(err)
	}
	_, ok := ndBC["http.send"]
	if !ok {
		t.Errorf("expected http.send cache entry")
	}
}
```
I've added this to `rego/rego_test.go` to illustrate the point (it fails). Should this not be included? After all, if the request had failed, that might be relevant for the later use of the decision log entries, wouldn't it? I mean, it's important even if the result isn't used for further evaluation...

Reply:
No, you're correct. I did a naive copy/paste/inversion of the conditional, and that clearly wasn't adequate. :slightly_frowning_face: 

Thank you for pointing out what I missed there. :sweat_smile: I'll push up the changes (and the new test case) later this afternoon.

Reply:
@srenatus That is an excellent example, and thank you for putting it together. I'll work in the new testcase, and fix the broken case(s) later today if possible.